### ISSUE TITLE
train_eval.py

### ISSUE THREAD

**Issue by karthik78437** [karthik78437/karthik#6]
"""
train_eval.py

Top-level script to train baseline LSTM and Attention-LSTM, evaluate and save results.
It does:
 - data generation
 - create sequences
 - train/test split
 - scaling
 - model creation, training
 - evaluation (RMSE, MAE, MAPE)
 - saving models and numpy predictions
"""

import os
import numpy as np
import json
from tensorflow.keras.callbacks import EarlyStopping, ReduceLROnPlateau, ModelCheckpoint

from data_generation import generate_complex_timeseries, create_sequences
from utils import train_test_split_sequences, scale_sequences, rmse, mape
from models import build_baseline_lstm, build_attention_lstm

def train_and_evaluate(seq_len=48,
                       test_size=0.2,
                       lstm_units=64,
                       dropout=0.2,
                       batch_size=64,
                       epochs=100,
                       results_dir='results'):
    os.makedirs(results_dir, exist_ok=True)

    print("Generating data...")
    df = generate_complex_timeseries(n_steps=2000, n_exog=3, seed=42)
    X, y = create_sequences(df, seq_len=seq_len)
    print("Sequences:", X.shape, y.shape)

    X_tr, X_te, y_tr, y_te = train_test_split_sequences(X, y, test_size=test_size)

    print("Scaling data...")
    X_tr_s, X_te_s, scaler = scale_sequences(X_tr, X_te, scaler_path=os.path.join(results_dir, 'scaler.joblib'))

    input_shape = (X_tr_s.shape[1], X_tr_s.shape[2])

    # Baseline LSTM
    print("Building baseline LSTM...")
    baseline = build_baseline_lstm(input_shape, lstm_units=lstm_units, dropout=dropout, bidirectional=False)
    baseline.summary()
    baseline_ckpt = os.path.join(results_dir, 'baseline.h5')
    cb_baseline = [
        EarlyStopping(monitor='val_loss', patience=10, restore_best_weights=True),
        ReduceLROnPlateau(monitor='val_loss', factor=0.5, patience=5),
        ModelCheckpoint(baseline_ckpt, save_best_only=True, monitor='val_loss')
    ]
    print("Training baseline LSTM...")
    baseline.fit(X_tr_s, y_tr, validation_split=0.1, epochs=epochs, batch_size=batch_size, callbacks=cb_baseline, verbose=2)
    baseline_preds = baseline.predict(X_te_s).flatten()
    baseline_metrics = {
        'rmse': float(rmse(y_te, baseline_preds)),
        'mae': float(np.mean(np.abs(y_te - baseline_preds))),
        'mape': float(mape(y_te, baseline_preds))
    }
    print("Baseline metrics:", baseline_metrics)
    baseline.save(baseline_ckpt)

    # Attention LSTM
    print("Building attention LSTM...")
    attention_model = build_attention_lstm(input_shape, lstm_units=lstm_units, dropout=dropout, attention_return_attention=False)
    attention_ckpt = os.path.join(results_dir, 'attention.h5')
    cb_attention = [
        EarlyStopping(monitor='val_loss', patience=12, restore_best_weights=True),
        ReduceLROnPlateau(monitor='val_loss', factor=0.5, patience=6),
        ModelCheckpoint(attention_ckpt, save_best_only=True, monitor='val_loss')
    ]
    print("Training attention LSTM...")
    attention_model.fit(X_tr_s, y_tr, validation_split=0.1, epochs=epochs, batch_size=batch_size, callbacks=cb_attention, verbose=2)
    attention_preds = attention_model.predict(X_te_s).flatten()
    attention_metrics = {
        'rmse': float(rmse(y_te, attention_preds)),
        'mae': float(np.mean(np.abs(y_te - attention_preds))),
        'mape': float(mape(y_te, attention_preds))
    }
    print("Attention metrics:", attention_metrics)
    attention_model.save(attention_ckpt)

    # Save predictions + metrics
    np.save(os.path.join(results_dir, 'y_test.npy'), y_te)
    np.save(os.path.join(results_dir, 'baseline_preds.npy'), baseline_preds)
    np.save(os.path.join(results_dir, 'attention_preds.npy'), attention_preds)

    with open(os.path.join(results_dir, 'metrics.json'), 'w') as f:
        json.dump({'baseline': baseline_metrics, 'attention': attention_metrics}, f, indent=2)

    return {'baseline': baseline_metrics, 'attention': attention_metrics}

if __name__ == '__main__':
    metrics = train_and_evaluate(seq_len=48, epochs=60, batch_size=128, lstm_units=64)
    print("Final metrics:", metrics)
________________________________________
